velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,585 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/sparksql/aggregates/ModeAggregate.h"
+
+#include "folly/container/F14Map.h"
+
+#include "velox/exec/AddressableNonNullValueList.h"
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/exec/Strings.h"
+
+using namespace facebook::velox::aggregate::prestosql;
+using namespace facebook::velox::exec;
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+namespace {
+
+struct UnknownValueHash {
```

Comment:
Since all nulls are skipped and the values map should not be accessed, do we need to define the hash for unknown type? Does it make sense to add some fast path for unknown type as its result is determined?

Reply:
This is just a placeholder, as we use simple aggregate adapter with default null behaviors for unknown values, all null inouts are ignored. I think this already covers the fast path youe mentiond, if not please let me know.
https://facebookincubator.github.io/velox/develop/aggregate-functions.html#default-null-behavior

Reply:
I think we can use folly::hasher directly, removed it.